orttraining/orttraining/python/training/torchdynamo/ort_backend.py

```diff
@@ -0,0 +1,519 @@
+
+
+def _create_onnx_proto(script_module):
+    onnx_proto = _jit_graph_to_onnx_model(script_module.graph, torch.onnx.OperatorExportTypes.ONNX)
+    return onnx_proto
+
+
+def _create_onnx_model(onnx_proto):
+    return onnx.ModelProto.FromString(onnx_proto)
+
+
+def _create_onnx_session(onnx_proto, device):
+    # TODO(wechi): Add more EPs per PyTorch device types.
+    # TODO(wechi): enable external allocators.
+    if device.type == "cuda":
+        return onnxruntime.InferenceSession(onnx_proto, providers=["CUDAExecutionProvider"])
+    return onnxruntime.InferenceSession(onnx_proto, providers=["CPUExecutionProvider"])
+
+
+def _get_onnx_devices(values: Tuple[torch.Tensor]) -> Tuple[ORTC.OrtDevice]:  # type: ignore
+    assert all(value.device == values[0].device for value in values), "All values must be on the same device."
+
+    def _device_id_or_zero(device_id: int) -> int:
+        return device_id or 0
+
+    devices: Tuple[ORTC.OrtDevice] = tuple(  # type: ignore
+        ORTC.OrtDevice(  # type: ignore
+            _get_ort_device_type(value.device.type, _device_id_or_zero(value.device.index)),
+            ORTC.OrtDevice.default_memory(),  # type: ignore
+            _device_id_or_zero(value.device.index),
+        )
+        for value in values
+    )
+    return devices
+
+
+def _run_onnx_session_with_ortvaluevector(
+    sess: onnxruntime.InferenceSession,
+    input_names: Tuple[str],
+    inputs: Tuple[torch.Tensor],
+    input_devices: Tuple[ORTC.OrtDevice],  # type: ignore
+    output_names: Tuple[str],
+    outputs: Tuple[torch.Tensor],
+    output_devices: Tuple[ORTC.OrtDevice],  # type: ignore
+):
+    _nvtx_range_push("contiguous")
+    inputs = tuple(a.contiguous() for a in inputs)
+    _nvtx_range_pop()
+
+    _nvtx_range_push("push_back_batch")
+    ort_inputs = ORTC.OrtValueVector()  # type: ignore
+    ort_inputs.reserve(len(inputs))
+    ort_outputs = ORTC.OrtValueVector()  # type: ignore
+    dtypes = []
+    shapes = []
+    data_ptrs = []
+
+    for value in inputs:
+        dtypes.append(_NP_DTYPE[value.dtype])
+        shapes.append(value.size())
+        data_ptrs.append(value.data_ptr())
+    ort_inputs.push_back_batch(inputs, data_ptrs, dtypes, shapes, input_devices)
+    _nvtx_range_pop()
+
+    _nvtx_range_push("run_with_ortvaluevector")
+    run_options = onnxruntime.RunOptions()
+    run_options.synchronize_execution_providers = True
+    sess.run_with_ortvaluevector(run_options, input_names, ort_inputs, output_names, ort_outputs, output_devices)
+    _nvtx_range_pop()
+
+    _nvtx_range_push("after run_with_ortvaluevector")
+    pth_outputs = onnxruntime.training.ortmodule._utils._ortvalues_to_torch_tensor(ort_outputs)  # type: ignore
+    _nvtx_range_pop()
+    return pth_outputs
+
+
+def _assert_allclose_with_detailed_error_message(
+    actual: torch.Tensor, expected: torch.Tensor, rtol: float = 1e-03, atol: float = 1e-04
+):
```

Comment:
Does `torch.testing.assert_close` work?

https://pytorch.org/docs/stable/testing.html#torch.testing.assert_close

Reply:
No. I need to compute atol and rtol. Those function just throws.